### PR TITLE
feat: Support Gemini via Vertex AI

### PR DESCRIPTION
**Pull request by dividedmind**
Note to access the API you have to create a service account at https://console.cloud.google.com/apis/credentials — after creation you'll get a JSON file with credentials. This JSON has to be put into `GOOGLE_WEB_CREDENTIALS` environment variable.

Example configuration:
```
GOOGLE_WEB_CREDENTIALS='{   "type": "service_account",   "project_id": "gen-lang-client-0551265112",   "private_key_id": "1fd731b64e9543d54e2921d7cfbe40c5f6f2845a",   "private_key": "-----BEGIN PRIVATE KEY-----\n...\n-----END PRIVATE KEY-----\n",   "client_email": "divide-s-navie-experiments@gen-lang-client-0551265112.iam.gserviceaccount.com",   "client_id": "112105442388554579010",   "auth_uri": "https://accounts.google.com/o/oauth2/auth",   "token_uri": "https://oauth2.googleapis.com/token",   "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",   "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/divide-s-navie-experiments%40gen-lang-client-0551265112.iam.gserviceaccount.com",   "universe_domain": "googleapis.com" }'
APPMAP_NAVIE_MODEL=gemini-1.5-pro-002
```

### CODE REVIEW

**Review by kgilpin — COMMENTED** [review]
Thanks! There are some build errors so I flipped this to Draft, temporarily.

File: packages/navie/package.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
   "dependencies": {
     "@langchain/anthropic": "^0.3.1",
     "@langchain/core": "^0.2.27",
+    "@langchain/google-vertexai-web": "^0.1.0",
```
Comment (dustinbyrne) [inline packages/navie/package.json:44]: Does this also work with the Gemini API described here?
https://ai.google.dev/gemini-api/docs/api-key

Note that they're referencing a `generativelanguage.googleapis.com` endpoint and not Vertex.

Comment (dividedmind) [reply packages/navie/package.json:44]: It's a different API: https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/overview#how-gemini-vertex-different-gemini-aistudio

Frankly, it's not clear to me which API we should use, but Langchain seems to only support Vertex.

Comment (kgilpin) [reply packages/navie/package.json:44]: Have we determined if the usage credits we have from Google will apply?

**Comment by kgilpin** [comment]
Tests still failing.

**Comment by dividedmind** [comment]
I forked the library that was ESM-only and added a CommonJS build (https://github.com/kriasoft/web-auth-library/pull/32) — hopefully this will make it work without fuss.

**Comment by kgilpin** [comment]
Thanks @dividedmind 

Do you have thoughts on the error we were seeing in the benchmark? Would really like to get a clean run though with Gemini.

**Comment by dividedmind** [comment]
> Thanks @dividedmind
> 
> Do you have thoughts on the error we were seeing in the benchmark? Would really like to get a clean run though with Gemini.

You mean the recitation problem? I'm looking at it next.

File: packages/navie/src/services/google-vertexai-completion-service.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        break;
+      } catch (cause) {
+        if (attempt < maxAttempts - 1 && tokens.length === 0) {
+          const nextAttempt = CompletionRetryDelay * 2 ** attempt;
+          warn(`Received ${JSON.stringify(cause)}, retrying in ${nextAttempt}ms`);
+          await new Promise<void>((resolve) => {
+            setTimeout(resolve, nextAttempt);
+          });
+          continue;
+        }
```
Comment (dustinbyrne) [inline packages/navie/src/services/google-vertexai-completion-service.ts:119]: Note that this doesn't attempt to prune input tokens

File: packages/navie/src/services/vector-terms-service.ts [hunk 3]
```diff
@@ -90,15 +85,20 @@ Terms: test cases +logContext jest`,
...
+    }),
     role: 'assistant',
   },
 ];
 
-const parseText = (text: string): string[] => text.split(/\s+/);
+const schema = z.object({
+  context: z.string(),
+  instructions: z.string(),
+  terms: z.array(z.string()),
```
Comment (dustinbyrne) [inline packages/navie/src/services/vector-terms-service.ts:100]: The description of these properties could move out of the prompt and in to the schema

**Review by dustinbyrne — APPROVED** [review]

File: packages/navie/src/services/google-vertexai-completion-service.ts (hunk 2 above)
Comment (dividedmind) [reply packages/navie/src/services/google-vertexai-completion-service.ts:119]: No, I figured it's unlikely to be necessary with 2M token input limit...

File: packages/navie/src/services/vector-terms-service.ts (hunk 3 above)
Comment (dividedmind) [reply packages/navie/src/services/vector-terms-service.ts:100]: True, although I tried to change it as little as possible as not to break anything.

**Pull request merged**